storage/blockchain/src/service/write.rs

```diff
@@ -55,7 +64,170 @@ fn write_block(env: &ConcreteEnv, block: &VerifiedBlockInformation) -> ResponseR
+        let mut last_block_height = 0;
+        for _ in 0..numb_blocks {
+            (last_block_height, _, _) =
+                crate::ops::block::pop_block(Some(old_main_chain_id), &mut tables_mut)?;
+        }
+
+        // Update the alt_chain_info with the correct information.
+        tables_mut.alt_chain_infos_mut().put(
+            &old_main_chain_id.into(),
+            &AltChainInfo {
+                parent_chain: Chain::Main.into(),
+                common_ancestor_height: last_block_height - 1,
+                chain_height: last_block_height + numb_blocks,
+            },
+        )?;
+
+        Ok(old_main_chain_id)
+    };
+
+    match result() {
+        Ok(old_main_chain_id) => {
+            TxRw::commit(tx_rw)?;
+            Ok(BlockchainResponse::PopBlocks(old_main_chain_id))
+        }
+        Err(e) => {
+            // INVARIANT: ensure database atomicity by aborting
+            // the transaction on `add_block()` failures.
+            TxRw::abort(tx_rw)
+                .expect("could not maintain database atomicity by aborting write transaction");
+            Err(e)
+        }
+    }
+}
+
+/// [`BlockchainWriteRequest::ReverseReorg`].
+fn reverse_reorg(env: &ConcreteEnv, chain_id: ChainId) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let mut tx_rw = env_inner.tx_rw()?;
+
+    // FIXME: turn this function into a try block once stable.
+    let mut result = || {
+        let mut tables_mut = env_inner.open_tables_mut(&tx_rw)?;
+
+        let chain_info = tables_mut.alt_chain_infos().get(&chain_id.into())?;
+        // Although this doesn't guarantee the chain was popped from the main-chain, it's an easy
+        // thing for us to check.
+        assert_eq!(Chain::from(chain_info.parent_chain), Chain::Main);
+
+        let top_block_height =
+            crate::ops::blockchain::top_block_height(tables_mut.block_heights())?;
+
+        // pop any blocks that were added as part of a re-org.
+        for _ in chain_info.common_ancestor_height..top_block_height {
+            crate::ops::block::pop_block(None, &mut tables_mut)?;
+        }
+
+        // Rust borrow rules requires us to collect into a Vec first before looping over the Vec.
+        let alt_blocks = ((chain_info.common_ancestor_height + 1)..chain_info.chain_height)
+            .map(|height| {
+                crate::ops::alt_block::get_alt_block(
+                    &AltBlockHeight {
+                        chain_id: chain_id.into(),
+                        height,
+                    },
+                    &tables_mut,
+                )
+            })
+            .collect::<Vec<_>>();
+
+        // Add the old main chain blocks back to the main chain.
+        for res_alt_block in alt_blocks {
+            let alt_block = res_alt_block?;
+
+            let verified_block = map_valid_alt_block_to_verified_block(alt_block);
+
+            crate::ops::block::add_block(&verified_block, &mut tables_mut)?;
+        }
```

Comment:
Is this comment true?
```suggestion
        // Add the old main chain blocks back to the main chain.
        for height in (chain_info.common_ancestor_height + 1)..chain_info.chain_height {
            let alt_block = crate::ops::alt_block::get_alt_block(
                &AltBlockHeight {
                    chain_id: chain_id.into(),
                    height,
                },
                &tables_mut,
            )?;

            let verified_block = map_valid_alt_block_to_verified_block(alt_block);

            crate::ops::block::add_block(&verified_block, &mut tables_mut)?;
        }
```

Reply:
oops, I retrieved the alt_block in an iterator, this is a better solution though